js/client-battle.js

```diff
@@ -1083,6 +1112,238 @@
 		},
 		hideTooltip: function() {
 			$('#tooltipwrapper').html('');
+		},
+		// Functions to calculate speed ranges of an opponent.
+		getTemplateMinSpeed: function (template, level) {
+			var nature = (this.battle.gen < 3 || this.battle.tier === 'Random Battle') ? 1 : 0.9;
+			return Math.floor(Math.floor(2 * template.baseStats['spe'] * level / 100 + 5) * nature);
+		},
+		getTemplateMaxSpeed: function (template, level) {
+			var nature = (this.battle.gen < 3 || this.battle.tier === 'Random Battle') ? 1 : 1.1;
+			var maxIvs = (this.battle.gen > 2) ? 31 : 30;
+			return Math.floor(Math.floor(Math.floor(2 * template.baseStats['spe'] + maxIvs + 63) * level / 100 + 5) * nature);
+		},
+		// Gets the proper current type for moves with a variable type.
+		getMoveType: function(move, pokemon) {
+			var moveType = move.type;
+			// Normalize is the first move type changing effect.
+			if (pokemon.ability === 'Normalize') {
+				moveType = 'Normal';
+			}
+			// Moves that require an item to change their type.
+			if (!this.battle.hasPseudoWeather('Magic Room') && !pokemon.volatiles['embargo']) {
+				if (move.id === 'judgment') {
+					var item = Tools.getItem(pokemon.item);
+					if (item.onPlate) moveType = item.onPlate;
+				}
+				if (move.id === 'technoblast') {
+					var item = Tools.getItem(pokemon.item);
+					if (item.onDrive) moveType = item.onDrive;
+				}
+				if (move.id === 'naturalgift' && pokemon.item) {
+					var item = Tools.getItem(pokemon.item);
+					if (item.naturalGift) moveType = item.naturalGift.type;
+				}
+			}
+			// Weather and pseudo-weather type changes.
+			if (move.id === 'naturepower') {
+				if (this.battle.hasPseudoWeather('Electric Terrain')) moveType = 'Electric';
+				if (this.battle.hasPseudoWeather('Grassy Terrain')) moveType = 'Grass';
+				if (this.battle.hasPseudoWeather('Misty Terrain')) moveType = 'Fairy';
+			}
```

Comment:
What about the case for Gen 5 Nature Power, which becomes Earthquake and therefore is Ground-type?